gwpy/timeseries/timeseries.py

```diff
@@ -1411,6 +1411,78 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
         phase.override_unit('deg' if deg else 'rad')
         return (mag, phase)
 
+    def heterodyne(self, phase, stride=1, exp=False, deg=True,
+                   singlesided=False):
+        """Compute the average magnitude and phase of this `TimeSeries`
```

Comment:
@mattpitkin, if I'm reading your original feature request correctly, you want this method to always return a complex number rather than a `(mag, phase)` tuple, correct? So I think you can clean up the use case and docstring:

```python
    def heterodyne(self, phase, stride=1, singlesided=False):
        """Compute the average magnitude and phase of this `TimeSeries`
        once per stride after heterodyning with a given phase series

        Parameters
        ----------
        phase : `array_like`
            an array of phase measurements (radians) with which to heterodyne
            the signal

        stride : `float`, optional
            stride (seconds) between calculations, defaults to 1 second

        singlesided : `bool`, optional
            Boolean switch to return single-sided output (i.e., to multiply by 2
            so that the signal is distributed across positive frequencies only),
            default: False

        Returns
        -------
        out : `TimeSeries`
            magnitude and phase trends, represented as
            `mag * exp(1j*phase)` with `dt=stride`

        Notes
        -----
        This is similar to the :meth:`~gwpy.timeseries.TimeSeries.demodulate`
        method, but is more general in that it accepts a varying phase evolution,
        rather than a fixed frequency.

        Unlike :meth:`~gwpy.timeseries.TimeSeries.demodulate`, the complex
        output is double-sided by default, so is not multiplied by 2.

        See also
        --------
        TimeSeries.demodulate
            for a method to heterodyne at a fixed frequency
        """
        try:
            phaselen = len(phase)
        except Exception as exc:
            raise TypeError("Phase series is not array_like: {}".format(exc))
        if phaselen != len(self):
            raise ValueError(
                "Phase array must be the same length as the TimeSeries"
            )
        stridesamp = int(stride * self.sample_rate.value)
        nsteps = int(self.size // stridesamp)
        # stride through the TimeSeries and heterodyne
        out = type(self)(numpy.zeros(nsteps, dtype=complex))
        out.__array_finalize__(self)
        out.sample_rate = 1 / float(stride)
        phasearray = numpy.asarray(phase)  # make sure phase is a numpy array
        for step in range(nsteps):
            istart = int(stridesamp * step)
            iend = istart + stridesamp
            idx = numpy.arange(istart, iend)
            mixed = numpy.exp(-1j * phasearray[idx]) * self.value[idx]
            out.value[step] = 2 * mixed.mean() if singlesided else mixed.mean()
        return out
```

Reply:
Thanks, your changes look good. I'd only ever be using it to output the complex time series rather than phases and magnitudes, and if anyone wants those quantities they are easy to generate.

Reply:
I've made the changes based on your suggestion